### ISSUE TITLE
The Eigen CPU backend's memory usage seems to scale with the number of threads.

### ISSUE THREAD

**Issue by OmnipotentEntity** [lightvector/KataGo#308]
I was running a benchmark on my PC, which has a lot of CPU cores, with the CPU version of katago (just to test it out and make sure that it compiles and runs fine), and the benchmark went through and automatically tripled the number of threads and tried again, as it does sometimes.  This wound up dragging my computer to a screeching halt as it used all of my working memory (64GB) and started tearing through my swap, so I had to drop to root text terminal to terminate the process.

Tested using an older 20block network that I just happened to have handy and the gtp_example.cfg.  Memory usage as reported via `ps`, in order of appearance.

threads | % of 64GB of memory
-|-
5 | 5.2%
12 | 9.6%
10 | 10.3% (leak?)

After which (this attempt) it crashed after throwing `std::bad_alloc`

```
$ katago benchmark -model g170-b20c256x2-s668214784-d222255714.txt.gz -config ../configs/gtp_example.cfg
2020-08-24 12:08:32-0400: Loading model and initializing benchmark...
2020-08-24 12:08:32-0400: nnRandSeed0 = 12874967738232564533
2020-08-24 12:08:32-0400: After dedups: nnModelFile0 = g170-b20c256x2-s668214784-d222255714.txt.gz useFP16 auto useNHWC auto
2020-08-24 12:08:41-0400: Eigen (CPU) backend thread 0: Model version 8
2020-08-24 12:08:41-0400: Eigen (CPU) backend thread 0: Model name: b20c256x2-s668214784-d222255714
2020-08-24 12:08:41-0400: Eigen (CPU) backend thread 2: Model version 8
2020-08-24 12:08:41-0400: Eigen (CPU) backend thread 2: Model name: b20c256x2-s668214784-d222255714
2020-08-24 12:08:41-0400: Eigen (CPU) backend thread 3: Model version 8
2020-08-24 12:08:41-0400: Eigen (CPU) backend thread 1: Model version 8
2020-08-24 12:08:41-0400: Eigen (CPU) backend thread 1: Model name: b20c256x2-s668214784-d222255714
2020-08-24 12:08:41-0400: Eigen (CPU) backend thread 4: Model version 8
2020-08-24 12:08:41-0400: Eigen (CPU) backend thread 3: Model name: b20c256x2-s668214784-d222255714
2020-08-24 12:08:41-0400: Eigen (CPU) backend thread 5: Model version 8
2020-08-24 12:08:41-0400: Eigen (CPU) backend thread 4: Model name: b20c256x2-s668214784-d222255714
2020-08-24 12:08:41-0400: Eigen (CPU) backend thread 5: Model name: b20c256x2-s668214784-d222255714

2020-08-24 12:08:45-0400: Loaded config ../configs/gtp_example.cfg
2020-08-24 12:08:45-0400: Loaded model g170-b20c256x2-s668214784-d222255714.txt.gz

Testing using 80 visits.
  If you have a good GPU, you might increase this using "-visits N" to get more accurate results.
  If you have a weak GPU and this is taking forever, you can decrease it instead to finish the benchmark faster.

You are currently using the Eigen (CPU) version of KataGo. Due to having no GPU, it may be slow.

Your GTP config is currently set to use numSearchThreads = 6
Automatically trying different numbers of threads to home in on the best: 

2020-08-24 12:08:45-0400: nnRandSeed0 = 903658743844004158
2020-08-24 12:08:45-0400: After dedups: nnModelFile0 = g170-b20c256x2-s668214784-d222255714.txt.gz useFP16 auto useNHWC auto
2020-08-24 12:08:53-0400: Eigen (CPU) backend thread 2: Model version 8
2020-08-24 12:08:53-0400: Eigen (CPU) backend thread 2: Model name: b20c256x2-s668214784-d222255714
2020-08-24 12:08:53-0400: Eigen (CPU) backend thread 3: Model version 8
2020-08-24 12:08:53-0400: Eigen (CPU) backend thread 0: Model version 8
2020-08-24 12:08:53-0400: Eigen (CPU) backend thread 1: Model version 8
2020-08-24 12:08:53-0400: Eigen (CPU) backend thread 1: Model name: b20c256x2-s668214784-d222255714
2020-08-24 12:08:53-0400: Eigen (CPU) backend thread 0: Model name: b20c256x2-s668214784-d222255714
2020-08-24 12:08:53-0400: Eigen (CPU) backend thread 3: Model name: b20c256x2-s668214784-d222255714
2020-08-24 12:08:53-0400: Eigen (CPU) backend thread 4: Model version 8
2020-08-24 12:08:53-0400: Eigen (CPU) backend thread 4: Model name: b20c256x2-s668214784-d222255714
2020-08-24 12:08:53-0400: Eigen (CPU) backend thread 5: Model version 8
2020-08-24 12:08:53-0400: Eigen (CPU) backend thread 5: Model name: b20c256x2-s668214784-d222255714


Possible numbers of threads to test: 1, 2, 3, 4, 5, 6, 8, 10, 12, 16, 20, 24, 32, 

numSearchThreads =  5: 10 / 10 positions, visits/s = 8.23 nnEvals/s = 8.15 nnBatches/s = 6.76 avgBatchSize = 1.21 (102.0 secs)
numSearchThreads = 12: 10 / 10 positions, visits/s = 16.21 nnEvals/s = 16.07 nnBatches/s = 12.20 avgBatchSize = 1.32 (56.1 secs)
numSearchThreads = 10: 10 / 10 positions, visits/s = 15.29 nnEvals/s = 15.05 nnBatches/s = 12.19 avgBatchSize = 1.23 (58.2 secs)
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
^C[1]    5450 abort (core dumped)  katago benchmark -model g170-b20c256x2-s668214784-d222255714.txt.gz -config
```

**Comment by lightvector** [comment]
No memory leak, just two instances of memory silliness.

https://github.com/lightvector/KataGo/pull/309

Does this branch work for you?

**Comment by OmnipotentEntity** [comment]
threads | %
-|-
5 | 1.4%
12 | 1.6%
10 | 1.7%
20 | 1.9%
16 | 2.1%
24 | 2.2%

Memory usage is much better.

**Comment by lightvector** [comment]
Released in v1.6.1. Yay, thanks.
https://github.com/lightvector/KataGo/releases
